infrastructure/src/main/java/org/corfudb/infrastructure/log/StreamLogFiles.java

```diff
@@ -146,6 +174,170 @@ public void sync() throws IOException {
+                trimLogFile(sh.getFileName(), pending);
+            } catch (IOException e) {
+                log.error("Compact operation failed for file {}", sh.getFileName());
+            }
+        }
+    }
+
+    private void trimLogFile(String filePath, Set<Long> pendingTrim) throws IOException {
+        FileChannel fc = FileChannel.open(FileSystems.getDefault().getPath(filePath + ".copy"),
+                EnumSet.of(StandardOpenOption.TRUNCATE_EXISTING, StandardOpenOption.WRITE,
+                        StandardOpenOption.CREATE, StandardOpenOption.SPARSE));
+
+        FileChannel fc2 = FileChannel.open(FileSystems.getDefault().getPath(getTrimmedFilePath(filePath)),
+                EnumSet.of(StandardOpenOption.APPEND));
+
+        Pair<LogHeader, List<LogEntry>> log = getCompactedEntries(filePath, pendingTrim);
+        LogHeader header = log.getKey();
+        List<LogEntry> compacted = log.getValue();
+
+        writeHeader(fc, header.getVersion(), header.getVerifyChecksum());
+
+        for (LogEntry entry : compacted) {
+
+            ByteBuffer record = getByteBufferWithMetaData(entry);
+            ByteBuffer recordBuf = ByteBuffer.allocate(Short.BYTES // Delimiter
+                    + record.capacity());
+
+            recordBuf.putShort(RECORD_DELIMITER);
+            recordBuf.put(record.array());
+            recordBuf.flip();
+
+            fc.write(recordBuf);
+        }
+
+        fc.force(true);
+        fc.close();
+
+        OutputStream outputStream = Channels.newOutputStream(fc2);
+
+        // Todo(Maithem) How do we verify that the compacted file is correct?
+        for (Long address : pendingTrim) {
+            TrimEntry entry = TrimEntry.newBuilder()
+                    .setChecksum(getChecksum(address))
+                    .setAddress(address)
+                    .build();
+            entry.writeDelimitedTo(outputStream);
+        }
+
+        outputStream.flush();
+        fc2.force(true);
+        fc2.close();
+
+        Files.move(Paths.get(filePath + ".copy"), Paths.get(filePath), StandardCopyOption.ATOMIC_MOVE);
+
+        // Force the reload of the new segment
+        writeChannels.remove(filePath);
+    }
+
+    Pair<LogHeader, List<LogEntry>> getCompactedEntries(String filePath, Set<Long> pendingTrim) throws IOException {
+        FileChannel fc = getChannel(filePath, true);
+
+        // Skip the header
+        ByteBuffer headerMetadataBuf = ByteBuffer.allocate(METADATA_SIZE);
+        fc.read(headerMetadataBuf);
+        headerMetadataBuf.flip();
+
+        Metadata headerMetadata = Metadata.parseFrom(headerMetadataBuf.array());
+        ByteBuffer headerBuf = ByteBuffer.allocate(headerMetadata.getLength());
+        fc.read(headerBuf);
+        headerBuf.flip();
+
+        LogHeader header = LogHeader.parseFrom(headerBuf.array());
+
+        ByteBuffer o = ByteBuffer.allocate((int) fc.size() - (int) fc.position());
```

Comment:
This is an optimization and its not obvious if its required or not, yet. The file size is kind of bounded by the hole filling timeout. The records per file can also be decreased to have more control over the total file size.